File: backend/game_loop.py
"""Game loop — ticks the world, processes agent state machines."""

from __future__ import annotations

import math
import random
from typing import Optional

from backend.state import (
    ActionType,
    Agent,
    AgentState,
    Building,
    GameWorld,
    Point,
    ResourceNode,
    ResourceType,
)

TICK_RATE = 2  # ticks per second


def tick_world(world: GameWorld, dt: float) -> list[dict]:
    """Advance the world by dt seconds. Returns events for broadcast."""
    world.tick_count += 1
    world.time_elapsed += dt
    events: list[dict] = []

    for agent in list(world.agents.values()):
        if agent.state == AgentState.DEAD:
            continue
        tick_agent(world, agent, dt, events)

    # Gentle resource regen
    for r in world.resources:
        if r.amount < r.max_amount:
            r.amount = min(r.max_amount, r.amount + 0.1 * dt)

    return events


# ── State Machine ────────────────────────────────────────────────────────
# Agents cycle: IDLE → (choose task) → ACTIVE → (task done) → IDLE
# No LLM needed yet — agents use simple heuristic rules.
# Later: swap in async LLM calls at the IDLE→choose-task transition.


def tick_agent(world: GameWorld, agent: Agent, dt: float, events: list[dict]) -> None:
    # DEAD or THINKING agents do nothing
    if agent.state in (AgentState.DEAD, AgentState.THINKING):
        return

    # IDLE → pick next action
    if agent.state == AgentState.IDLE:
        _choose_next_action(world, agent)
        if agent.current_action is None:
            # Nothing to do — stay idle
            return
        agent.action_progress = 0.0
        agent.state = AgentState.ACTIVE

    # ACTIVE → execute current action
    if agent.state == AgentState.ACTIVE and agent.current_action:
        action = agent.current_action
        action.progress += dt
        _execute_action(world, agent, action, dt, events)

        # Check completion
        if action.progress >= action.duration_seconds:
            on_action_done(world, agent, action, events)
            agent.current_action = None
            agent.action_progress = 0.0
            agent.state = AgentState.IDLE


def on_action_done(world: GameWorld, agent: Agent, action, events: list[dict]) -> None:
    """Side effects when an action completes."""
    pass  # post-completion effects handled inline per action


def _choose_next_action(world: GameWorld, agent: Agent) -> None:
    """Heuristic decision: pick what the agent does next based on state."""
    # Priority 1: deposit if carrying lots
    if agent.carry_weight() > 40:
        tc = _nearest_town_center(world, agent)
        if tc:
            agent.current_action = world._make_action(
                ActionType.DEPOSIT,
                duration_seconds=agent.position.dist(tc.position) / agent.speed + 1,
            )
            return

    # Priority 2: retreat if low health
    if agent.health < 25:
        tc = _nearest_town_center(world, agent)
        if tc:
            agent.current_action = world._make_action(
                ActionType.CAMP,
                duration_seconds=5.0,
            )
            return

    # Priority 3: gather nearest resource
    nearest = _nearest_resource(world, agent)
    if nearest:
        travel_time = agent.position.dist(nearest.position) / agent.speed
        gather_time = 8.0
        agent.current_action = world._make_action(
            ActionType.GATHER,
            target_id=nearest.id,
            duration_seconds=travel_time + gather_time,
        )
        return

    # Priority 4: wander
    wander_target = Point(
        random.uniform(50, world.width - 50),
        random.uniform(50, world.height - 50),
    )
    agent.current_action = world._make_action(
        ActionType.WANDER,
        target_position=wander_target,
        duration_seconds=agent.position.dist(wander_target) / agent.speed,
    )


def _execute_action(world: GameWorld, agent: Agent, action, dt: float, events: list[dict]) -> None:
    if action.action_type == ActionType.MOVE_TO:
        _exec_move(agent, action, dt)
    elif action.action_type == ActionType.WANDER:
        _exec_wander(agent, action, dt)
    elif action.action_type == ActionType.GATHER:
        _exec_gather(world, agent, action, dt, events)
    elif action.action_type == ActionType.DEPOSIT:
        _exec_deposit(world, agent, action, dt, events)
    elif action.action_type == ActionType.CAMP:
        _exec_camp(agent, action, dt)
    elif action.action_type == ActionType.BUILD:
        _exec_build(world, agent, action, dt, events)
    elif action.action_type == ActionType.IDLE:
        pass  # just wait


def _exec_move(agent: Agent, action, dt: float) -> None:
    if not action.target_position:
        action.progress = action.duration_seconds
        return
    _move_toward(agent, action.target_position, dt, action.duration_seconds)


def _exec_wander(agent: Agent, action, dt: float) -> None:
    if action.target_position:
        _move_toward(agent, action.target_position, dt, action.duration_seconds)


def _exec_gather(world: GameWorld, agent: Agent, action, dt: float, events: list[dict]) -> None:
    target = _find_resource(world, action.target_id)
    if target is None or not target.alive:
        # Resource depleted — find another
        new_target = _nearest_resource(world, agent)
        if new_target:
            action.target_id = new_target.id
            # Extend duration to account for travel
            travel = agent.position.dist(new_target.position) / agent.speed
            action.duration_seconds = action.progress + travel + 8.0
        else:
            action.progress = action.duration_seconds
        return

    # Move to resource if far
    if agent.position.dist(target.position) > 25:
        _move_toward(agent, target.position, dt, action.duration_seconds)
        return

    # Gather
    rate = 12.0 * dt
    gathered = min(rate, target.amount)
    target.amount -= gathered
    setattr(agent, target.kind.value, getattr(agent, target.kind.value, 0) + gathered)
    events.append({"type": "gather", "agent_id": agent.id, "resource": target.kind.value, "amount": round(gathered, 1)})

    if target.amount <= 0:
        action.target_id = None  # will repick next tick


def _exec_deposit(world: GameWorld, agent: Agent, action, dt: float, events: list[dict]) -> None:
    tc = _nearest_town_center(world, agent)
    if not tc:
        action.progress = action.duration_seconds
        return

    if agent.position.dist(tc.position) > 20:
        _move_toward(agent, tc.position, dt, action.duration_seconds)
        return

    # Deposit all carried resources
    total = 0
    for res in ("wood", "gold", "stone", "food"):
        val = getattr(agent, res, 0)
        total += val
        setattr(agent, res, 0)
    if total > 0:
        events.append({"type": "deposited", "agent_id": agent.id, "amount": round(total, 1)})
    action.progress = action.duration_seconds


def _exec_camp(agent: Agent, action, dt: float) -> None:
    agent.health = min(100.0, agent.health + 15.0 * dt)


def _exec_build(world: GameWorld, agent: Agent, action, dt: float, events: list[dict]) -> None:
    if not action.target_position:
        action.progress = action.duration_seconds
        return

    if agent.position.dist(action.target_position) > 20:
        _move_toward(agent, action.target_position, dt, action.duration_seconds)
        return

    # Spend resources and place building
    if agent.wood >= 50:
        agent.wood -= 50
        world.buildings.append(Building(
            kind=ResourceType.TOWN_CENTER,
            position=action.target_position,
            owner=agent.id,
        ))
        events.append({"type": "built", "agent_id": agent.id})
    action.progress = action.duration_seconds


# ── Movement helper ─────────────────────────────────────────────────────
def _move_toward(agent: Agent, target: Point, dt: float, timeout: float) -> None:
    dx = target.x - agent.position.x
    dy = target.y - agent.position.y
    dist = math.sqrt(dx * dx + dy * dy)

    if dist < 5:
        agent.position.x, agent.position.y = target.x, target.y
        return

    step = agent.speed * dt
    if step >= dist:
        agent.position.x, agent.position.y = target.x, target.y
    else:
        agent.position.x += (dx / dist) * step
        agent.position.y += (dy / dist) * step


# ── World queries ────────────────────────────────────────────────────────
def _nearest_town_center(world: GameWorld, agent: Agent) -> Optional[Point]:
    tcs = [b for b in world.buildings if b.kind == ResourceType.TOWN_CENTER]
    if not tcs:
        return None
    return min(tcs, key=lambda b: agent.position.dist(b.position)).position


def _nearest_resource(world: GameWorld, agent: Agent) -> Optional[ResourceNode]:
    alive = [r for r in world.resources if r.alive]
    if not alive:
        return None
    return min(alive, key=lambda r: agent.position.dist(r.position))


def _find_resource(world: GameWorld, res_id: Optional[str]) -> Optional[ResourceNode]:
    if res_id is None:
        return None
    for r in world.resources:
        if r.id == res_id:
            return r
    return None

File: src/game_loop.rs
/// Game loop — 2 Hz tick with agent state machine.

use rand::Rng;
use uuid::Uuid;

use crate::state::*;

pub const TICK_RATE: f64 = 2.0;

/// Advance the world by dt seconds. Returns events for broadcast.
pub fn tick_world(world: &mut GameWorld, dt: f64) -> Vec<WorldEvent> {
    world.tick_count += 1;
    world.time_elapsed += dt;
    let mut events: Vec<WorldEvent> = Vec::new();

    let agent_ids: Vec<String> = world.agents.keys().cloned().collect();
    for id in agent_ids {
        let is_dead = world
            .agents
            .get(&id)
            .is_some_and(|a| a.state == AgentState::Dead);
        if is_dead {
            continue;
        }
        tick_agent(world, &id, dt, &mut events);
    }

    // Gentle resource regen
    for r in &mut world.resources {
        if r.amount < r.max_amount {
            r.amount = (r.max_amount).min(r.amount + 0.1 * dt);
        }
    }

    events
}

fn tick_agent(world: &mut GameWorld, agent_id: &str, dt: f64, events: &mut Vec<WorldEvent>) {
    // Check state — bail if THINKING or DEAD
    {
        let agent = world.agents.get(agent_id).unwrap();
        if agent.state == AgentState::Dead || agent.state == AgentState::Thinking {
            return;
        }
    }

    // IDLE → choose an action
    if world.agents[agent_id].state == AgentState::Idle {
        choose_next_action(world, agent_id);
        let has_action = world.agents[agent_id].current_action.is_some();
        if !has_action {
            return;
        }
        // Set progress and go ACTIVE
        let agent = world.agents.get_mut(agent_id).unwrap();
        if let Some(ref mut action) = agent.current_action {
            action.progress = 0.0;
        }
        agent.state = AgentState::Active;
    }

    // ACTIVE → tick the current action
    if world.agents[agent_id].state == AgentState::Active {
        // Execute the current action logic
        run_execute_action(world, agent_id, dt, events);

        // Advance progress and check completion
        let finished = {
            let agent = world.agents.get_mut(agent_id).unwrap();
            if let Some(ref mut action) = agent.current_action {
                action.progress += dt;
                action.progress >= action.duration_seconds
            } else {
                true
            }
        };

        if finished {
            world.agents.get_mut(agent_id).unwrap().current_action = None;
            world.agents.get_mut(agent_id).unwrap().state = AgentState::Idle;
        }
    }
}

/// After tick_agent advances progress, execute_action is called separately
/// with the action already borrowed from world. This avoids NLL issues.
pub fn run_execute_action(world: &mut GameWorld, agent_id: &str, dt: f64, events: &mut Vec<WorldEvent>) {
    // Only execute when agent is Active and has an action
    let action_type = {
        let agent = match world.agents.get(agent_id) {
            Some(a) => a,
            None => return,
        };
        if agent.state != AgentState::Active {
            return;
        }
        let action = match &agent.current_action {
            Some(a) => a.action_type,
            None => return,
        };
        action
    };

    match action_type {
        ActionType::MoveTo | ActionType::Wander | ActionType::Scout => {
            exec_move(world, agent_id, dt);
        }
        ActionType::Gather => {
            exec_gather(world, agent_id, dt, events);
        }
        ActionType::Deposit => {
            exec_deposit(world, agent_id, dt, events);
        }
        ActionType::Camp => {
            exec_camp(world, agent_id, dt);
        }
        ActionType::Build => {
            exec_build(world, agent_id, dt, events);
        }
        ActionType::Attack => {
            exec_move(world, agent_id, dt);
        }
        ActionType::Idle => {}
    }
}

// ── Action selection (state machine) ─────────────────────────────────────

fn choose_next_action(world: &mut GameWorld, agent_id: &str) {
    let agent = world.agents.get(agent_id).unwrap().clone();

    // Priority 1: deposit if carrying lots
    if agent.carry_weight() > 40.0 {
        if let Some(tc_pos) = nearest_town_center(world, &agent) {
            let travel_time = agent.position.dist(&tc_pos) / agent.speed;
            let a = world.agents.get_mut(agent_id).unwrap();
            a.current_action = Some(Box::new(QueuedAction::new(
                ActionType::Deposit,
                None,
                Some(tc_pos),
                travel_time + 1.0,
            )));
            return;
        }
    }

    // Priority 2: camp if low health
    if agent.health < 25.0 {
        let a = world.agents.get_mut(agent_id).unwrap();
        a.current_action = Some(Box::new(QueuedAction::new(
            ActionType::Camp,
            None,
            None,
            5.0,
        )));
        return;
    }

    // Priority 3: gather nearest resource
    if let Some(resource) = nearest_resource(world, &agent) {
        let travel_time = agent.position.dist(&resource.position) / agent.speed;
        let res_id = resource.id.clone();  // clone before mutable borrow
        let a = world.agents.get_mut(agent_id).unwrap();
        a.current_action = Some(Box::new(QueuedAction::new(
            ActionType::Gather,
            Some(res_id),
            None,
            travel_time + 8.0,
        )));
        return;
    }

    // Priority 4: wander
    let mut rng = rand::thread_rng();
    let wander_target = Point::new(
        rng.gen_range(50.0..world.width - 50.0),
        rng.gen_range(50.0..world.height - 50.0),
    );
    let travel_time = agent.position.dist(&wander_target) / agent.speed;
    let a = world.agents.get_mut(agent_id).unwrap();
    a.current_action = Some(Box::new(QueuedAction::new(
        ActionType::Wander,
        None,
        Some(wander_target),
        travel_time,
    )));
}

// ── Movement ────────────────────────────────────────────────────────────

fn exec_move(world: &mut GameWorld, agent_id: &str, dt: f64) {
    let target = {
        let agent = world.agents.get(agent_id).unwrap();
        match &agent.current_action {
            Some(a) => match &a.target_position {
                Some(p) => *p,
                None => return,
            },
            None => return,
        }
    };

    let agent = world.agents.get_mut(agent_id).unwrap();
    let dx = target.x - agent.position.x;
    let dy = target.y - agent.position.y;
    let dist = (dx * dx + dy * dy).sqrt();

    if dist < 5.0 {
        agent.position.x = target.x;
        agent.position.y = target.y;
        if let Some(ref mut action) = agent.current_action {
            action.progress = action.duration_seconds;
        }
        return;
    }

    let step = agent.speed * dt;
    if step >= dist {
        agent.position.x = target.x;
        agent.position.y = target.y;
    } else {
        agent.position.x += (dx / dist) * step;
        agent.position.y += (dy / dist) * step;
    }
}

// ── Gathering ───────────────────────────────────────────────────────────

fn exec_gather(world: &mut GameWorld, agent_id: &str, dt: f64, events: &mut Vec<WorldEvent>) {
    let target_id = {
        let agent = world.agents.get(agent_id).unwrap();
        match &agent.current_action {
            Some(a) => a.target_id.clone(),
            None => return,
        }
    };

    let target_id = match target_id {
        Some(id) => id,
        None => return,
    };

    // Find resource index
    let res_idx = world.resources.iter().position(|r| r.id == target_id && r.alive());

    let res_idx = match res_idx {
        Some(idx) => idx,
        None => {
            // Resource depleted — agent will repick on next idle cycle
            // Just let this action finish
            let agent = world.agents.get_mut(agent_id).unwrap();
            if let Some(ref mut action) = agent.current_action {
                action.progress = action.duration_seconds;
            }
            return;
        }
    };

    // Move to resource if far
    {
        let agent = world.agents.get(agent_id).unwrap();
        let res = &world.resources[res_idx];
        if agent.position.dist(&res.position) > 25.0 {
            let _ = agent;
            let agent = world.agents.get_mut(agent_id).unwrap();
            let res = &world.resources[res_idx];
            let dx = res.position.x - agent.position.x;
            let dy = res.position.y - agent.position.y;
            let dist = (dx * dx + dy * dy).sqrt();
            let step = agent.speed * dt;
            if step >= dist {
                agent.position.x = res.position.x;
                agent.position.y = res.position.y;
            } else {
                agent.position.x += (dx / dist) * step;
                agent.position.y += (dy / dist) * step;
            }
            return;
        }
    }

    // Gather
    let agent = world.agents.get_mut(agent_id).unwrap();
    let resource = &mut world.resources[res_idx];
    let rate = 12.0 * dt;
    let gathered = rate.min(resource.amount);
    resource.amount -= gathered;

    match resource.kind {
        ResourceKind::Wood => agent.wood += gathered,
        ResourceKind::Gold => agent.gold += gathered,
        ResourceKind::Food => agent.food += gathered,
        ResourceKind::Stone => agent.stone += gathered,
    }

    events.push(WorldEvent {
        event_type: "gather".to_string(),
        agent_id: Some(agent_id.to_string()),
        resource: Some(format!("{:?}", resource.kind).to_lowercase()),
        amount: Some((gathered * 10.0).round() / 10.0),
    });
}

// ── Deposit ─────────────────────────────────────────────────────────────

fn exec_deposit(world: &mut GameWorld, agent_id: &str, dt: f64, events: &mut Vec<WorldEvent>) {
    let tc_pos = {
        let agent = world.agents.get(agent_id).unwrap();
        nearest_town_center(world, agent)
    };

    let tc_pos = match tc_pos {
        Some(p) => p,
        None => {
            let agent = world.agents.get_mut(agent_id).unwrap();
            if let Some(ref mut action) = agent.current_action {
                action.progress = action.duration_seconds;
            }
            return;
        }
    };

    // Move to town center if far
    {
        let agent = world.agents.get(agent_id).unwrap();
        if agent.position.dist(&tc_pos) > 20.0 {
            let _ = agent;
            let agent = world.agents.get_mut(agent_id).unwrap();
            let dx = tc_pos.x - agent.position.x;
            let dy = tc_pos.y - agent.position.y;
            let dist = (dx * dx + dy * dy).sqrt();
            let step = agent.speed * dt;
            if step >= dist {
                agent.position.x = tc_pos.x;
                agent.position.y = tc_pos.y;
            } else {
                agent.position.x += (dx / dist) * step;
                agent.position.y += (dy / dist) * step;
            }
            return;
        }
    }

    // Deposit
    let agent = world.agents.get_mut(agent_id).unwrap();
    let total = agent.wood + agent.gold + agent.stone + agent.food;
    agent.wood = 0.0;
    agent.gold = 0.0;
    agent.stone = 0.0;
    agent.food = 0.0;

    if total > 0.0 {
        events.push(WorldEvent {
            event_type: "deposited".to_string(),
            agent_id: Some(agent_id.to_string()),
            resource: None,
            amount: Some((total * 10.0).round() / 10.0),
        });
    }

    if let Some(ref mut action) = world.agents.get_mut(agent_id).unwrap().current_action {
        action.progress = action.duration_seconds;
    }
}

// ── Camp (heal) ─────────────────────────────────────────────────────────

fn exec_camp(world: &mut GameWorld, agent_id: &str, dt: f64) {
    let agent = world.agents.get_mut(agent_id).unwrap();
    agent.health = 100.0f64.min(agent.health + 15.0 * dt);
}

// ── Build ───────────────────────────────────────────────────────────────

fn exec_build(world: &mut GameWorld, agent_id: &str, dt: f64, events: &mut Vec<WorldEvent>) {
    let target = {
        let agent = world.agents.get(agent_id).unwrap();
        match &agent.current_action {
            Some(a) => a.target_position,
            None => return,
        }
    };

    let target = match target {
        Some(p) => p,
        None => return,
    };

    // Move to build site
    {
        let agent = world.agents.get(agent_id).unwrap();
        if agent.position.dist(&target) > 20.0 {
            let _ = agent;
            let agent = world.agents.get_mut(agent_id).unwrap();
            let dx = target.x - agent.position.x;
            let dy = target.y - agent.position.y;
            let dist = (dx * dx + dy * dy).sqrt();
            let step = agent.speed * dt;
            if step >= dist {
                agent.position.x = target.x;
                agent.position.y = target.y;
            } else {
                agent.position.x += (dx / dist) * step;
                agent.position.y += (dy / dist) * step;
            }
            return;
        }
    }

    // Build
    let agent = world.agents.get_mut(agent_id).unwrap();
    if agent.wood >= 50.0 {
        agent.wood -= 50.0;
        let building = Building {
            id: Uuid::new_v4().to_string()[..8].to_string(),
            kind: "town_center".to_string(),
            position: target,
            health: 100.0,
            owner: agent.id.clone(),
        };
        world.buildings.push(building);
        events.push(WorldEvent {
            event_type: "built".to_string(),
            agent_id: Some(agent_id.to_string()),
            resource: None,
            amount: None,
        });
    }

    if let Some(ref mut action) = world.agents.get_mut(agent_id).unwrap().current_action {
        action.progress = action.duration_seconds;
    }
}

// ── World queries ───────────────────────────────────────────────────────

fn nearest_town_center(world: &GameWorld, agent: &Agent) -> Option<Point> {
    world
        .buildings
        .iter()
        .filter(|b| b.kind == "town_center")
        .min_by(|a, b| {
            let da = agent.position.dist(&a.position);
            let db = agent.position.dist(&b.position);
            da.partial_cmp(&db).unwrap()
        })
        .map(|b| b.position)
}

fn nearest_resource<'a>(world: &'a GameWorld, agent: &Agent) -> Option<&'a ResourceNode> {
    world
        .resources
        .iter()
        .filter(|r| r.alive())
        .min_by(|a, b| {
            let da = agent.position.dist(&a.position);
            let db = agent.position.dist(&b.position);
            da.partial_cmp(&db).unwrap()
        })
}